c++/src/kj/one-of.h

```diff
@@ -67,18 +89,43 @@ using Variants = typename Variants_<i>::Type;
 template <typename... Variants>
 class OneOf {
   template <typename Key>
-  static inline constexpr uint typeIndex() { return _::TypeIndex_<1, Key, Variants...>::value; }
-  // Get the 1-based index of Key within the type list Types.
+  static inline constexpr uint typeIndex() {
+    return _::TypeIndex_<1, _::OneOfFailError_, Key, Variants...>::value;
+  }
+  // Get the 1-based index of Key within the type list Types, or static_assert with a nice error.
+
+  template <typename Key>
+  static inline constexpr uint typeIndexOrZero() {
+    return _::TypeIndex_<1, _::OneOfFailZero_, Key, Variants...>::value;
+  }
+
+  template <uint i, typename... OtherVariants>
+  struct HasAll;
+  // Has a member type called "Success" if and only if all of `OtherVariants` are types that
+  // appear in `Variants`. Used with SFINAE to enable subset constructors.
 
 public:
   inline OneOf(): tag(0) {}
+
   OneOf(const OneOf& other) { copyFrom(other); }
   OneOf(OneOf& other) { copyFrom(other); }
   OneOf(OneOf&& other) { moveFrom(other); }
-  template <typename T>
+  // Copy/move from same OneOf type.
+
+  template <typename... OtherVariants, typename = typename HasAll<1, OtherVariants...>::Success>
+  OneOf(const OneOf<OtherVariants...>& other) { copyFromSubset(other); }
+  template <typename... OtherVariants, typename = typename HasAll<1, OtherVariants...>::Success>
+  OneOf(OneOf<OtherVariants...>& other) { copyFromSubset(other); }
+  template <typename... OtherVariants, typename = typename HasAll<1, OtherVariants...>::Success>
```

Comment:
I'm curious if there's a significance to the `1` constant, versus the `0` on line 123? Since empty OneOfs (i.e. `OneOf<>`) aren't allowed, it seems like we'll always instantiate the `HasAll<i, First, Rest...>` specialization, rather than the `HasAll<i>` specialization, thus ignoring both the `0` and `1`.

Reply:
You thought about it harder than I did. I was just shoving garbage in there.